src/test/java/org/jenkinsci/plugins/envinject/EnvInjectPluginActionTest.java

```diff
@@ -0,0 +1,155 @@
+        plugin.configure(new EnvInjectPluginConfiguration(true, false));
+        
+        // Run build and retrieve results
+        FreeStyleBuild build = j.buildAndAssertSuccess(p);
+        assertEquals("BAR", capture.getEnvVars().get("FOO"));
+        assertEquals("BAR", build.getEnvironment(TaskListener.NULL).get("FOO"));
+        
+        // Check action
+        EnvInjectPluginAction action = build.getAction(EnvInjectPluginAction.class);
+        assertNotNull("EnvInjectPluginAction has not been injected", action);
+        Object target = action.getTarget();
+        assertNull(action.getIconFileName());
+        assertNull(action.getUrlName());
+        
+        assertEquals("Injected variables should be hidden", EnvInjectVarList.HIDDEN, target);
+        assertFalse("Users should have no permission to see injected vars", 
+                canViewInjectedVars(User.getUnknown(), build));
+    }
+    
+    @Test
+    public void hideActionViaPermissions() throws Exception {
+        // Enable permissions
+        final EnvInjectPlugin plugin = EnvInjectPlugin.getInstance();
+        plugin.configure(new EnvInjectPluginConfiguration(false, true));
+        
+        // Create a test user
+        User user = User.get("testUser", true, null);
+             
+        final GlobalMatrixAuthorizationStrategy strategy = new GlobalMatrixAuthorizationStrategy();
+        strategy.add(Jenkins.READ, user.getId());
+        strategy.add(Item.READ, user.getId());
+        strategy.add(Item.DISCOVER, user.getId());
+        j.jenkins.setAuthorizationStrategy(strategy);
+        
+        // Run build and retrieve results
+        FreeStyleBuild build = j.buildAndAssertSuccess(p);
+        assertFalse("User should have no View Injected Vars permission", 
+                hasPermission(user, build, EnvInjectPlugin.VIEW_INJECTED_VARS));
+        assertFalse("User should have no permission to see injected vars", 
+                canViewInjectedVars(user, build));
+            
+        // Grant permissions and check the results 
+        strategy.add(EnvInjectPlugin.VIEW_INJECTED_VARS, user.getId());
+        assertTrue("User should have the View Injected Vars permission", 
+                hasPermission(user, build, EnvInjectPlugin.VIEW_INJECTED_VARS));
+        assertTrue("User should have a permission to see injected vars", 
+                canViewInjectedVars(user, build));
+    }
+    
+    
+    private boolean hasPermission(User user, AccessControlled item, Permission permission ) 
+            throws AssertionError {
+        SecurityContext initialContext = null;
+        try {
+            Authentication auth = user.impersonate();
+            initialContext = hudson.security.ACL.impersonate(auth);
+            return item.hasPermission(permission);
+        } catch (UsernameNotFoundException ex) {
+            throw new AssertionError("The specified user is not found", ex);
```

Comment:
fail(...) does not allow to attach the nested exception. In any case, it just wraps the throw call, so I don't see an issue with the current approach